packages/graphql/src/schema/validation/custom-rules/valid-types/valid-list-in-node-type.ts

```diff
@@ -0,0 +1,116 @@
+export function ValidListInNodeType(context: SDLValidationContext): ASTVisitor {
+    return {
+        ObjectTypeDefinition(objectTypeDefinitionNode: ObjectTypeDefinitionNode, _key, _parent) {
+            const { directives } = objectTypeDefinitionNode;
+
+            const nodeUsage = directives?.find((directive) => directive.name.value === nodeDirective.name);
+            const relationshipPropertiesUsage = directives?.find(
+                (directive) => directive.name.value === relationshipPropertiesDirective.name
+            );
+            if (!directives || (!nodeUsage && !relationshipPropertiesUsage)) {
+                return; // Skip non-node types or relationshipProperties types
+            }
+
+            objectTypeDefinitionNode.fields?.forEach((fieldDefinitionNode) => {
+                const { type, directives } = fieldDefinitionNode;
+                if (
+                    directives &&
+                    directives.some(
+                        (directive) =>
+                            directive.name.value === cypherDirective.name ||
+                            directive.name.value === relationshipDirective.name
+                    )
+                ) {
+                    return; // Skip cypher fields and relationship fields, relationship fields have their own validation
+                }
+                const { isValid, errorMsg, errorPath } = assertValid(() => {
+                    const typePath = getTypePath(type);
+                    if (typePath.includes(Kind.LIST_TYPE)) {
+                        const wrappedType = getInnerTypeName(type);
+                        const validTypePaths: TypePath[] = [
+                            [Kind.LIST_TYPE, Kind.NON_NULL_TYPE, wrappedType],
+                            [Kind.NON_NULL_TYPE, Kind.LIST_TYPE, Kind.NON_NULL_TYPE, wrappedType],
+                        ];
+                        if (!findTypePathInTypePaths(typePath, validTypePaths)) {
+                            const typeStr = typeNodeToString(type);
+
+                            const directiveName = (nodeUsage ?? relationshipPropertiesUsage)?.name?.value;
+                            throw new DocumentValidationError(
+                                `List of nullable elements are not supported in "@${directiveName}" types. Found: ${typeStr}`,
+                                []
+                            );
+                        }
+                    }
+                });
+
+                if (!isValid) {
+                    context.reportError(
+                        createGraphQLError({
+                            nodes: [fieldDefinitionNode],
+                            path: [objectTypeDefinitionNode.name.value, fieldDefinitionNode.name.value, ...errorPath],
+                            errorMsg,
+                        })
+                    );
+                }
+            });
+        },
+    };
+}
+
+type TypePath = (string | Kind.LIST_TYPE | Kind.NON_NULL_TYPE)[];
```

Comment:
Nitpick (no need to change it) I find the array syntax less confusing when dealing with nested types:

```suggestion
type TypePath = Array<string | Kind.LIST_TYPE | Kind.NON_NULL_TYPE>;
```